src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonTypeInfo.cs

```diff
@@ -0,0 +1,695 @@
+        {
+            JsonTypeInfo untyped = typeInfo;
+            Assert.Equal(typeof(T), typeInfo.Type);
+            Assert.True(typeInfo.Converter.CanConvert(typeof(T)));
+
+            JsonPropertyInfo prop = typeInfo.CreateJsonPropertyInfo(typeof(string), "foo");
+            Assert.Throws<InvalidOperationException>(() => untyped.CreateObject = untyped.CreateObject);
+            Assert.Throws<InvalidOperationException>(() => typeInfo.CreateObject = typeInfo.CreateObject);
+            Assert.Throws<InvalidOperationException>(() => typeInfo.NumberHandling = typeInfo.NumberHandling);
+            Assert.Throws<InvalidOperationException>(() => typeInfo.Properties.Clear());
+            Assert.Throws<InvalidOperationException>(() => typeInfo.Properties.Add(prop));
+            Assert.Throws<InvalidOperationException>(() => typeInfo.Properties.Insert(0, prop));
+
+            foreach (var property in typeInfo.Properties)
+            {
+                Assert.NotNull(property.PropertyType);
+                Assert.Null(property.CustomConverter);
+                Assert.NotNull(property.Name);
+                Assert.NotNull(property.Get);
+                Assert.NotNull(property.Set);
+                Assert.Null(property.ShouldSerialize);
+                Assert.Null(typeInfo.NumberHandling);
+
+                Assert.Throws<InvalidOperationException>(() => property.CustomConverter = property.CustomConverter);
+                Assert.Throws<InvalidOperationException>(() => property.Name = property.Name);
+                Assert.Throws<InvalidOperationException>(() => property.Get = property.Get);
+                Assert.Throws<InvalidOperationException>(() => property.Set = property.Set);
+                Assert.Throws<InvalidOperationException>(() => property.ShouldSerialize = property.ShouldSerialize);
+                Assert.Throws<InvalidOperationException>(() => property.NumberHandling = property.NumberHandling);
+            }
+        }
+
+        [Theory]
+        [InlineData(typeof(object), JsonTypeInfoKind.None)]
+        [InlineData(typeof(string), JsonTypeInfoKind.None)]
+        [InlineData(typeof(int), JsonTypeInfoKind.None)]
+        [InlineData(typeof(SomeRecursiveClass) /* custom converter */, JsonTypeInfoKind.None)]
+        [InlineData(typeof(SomeClass), JsonTypeInfoKind.Object)]
+        [InlineData(typeof(DefaultJsonTypeInfoResolverTests), JsonTypeInfoKind.Object)]
+        [InlineData(typeof(StructWithFourArgs), JsonTypeInfoKind.Object)]
+        [InlineData(typeof(Dictionary<string, string>), JsonTypeInfoKind.Dictionary)]
+        [InlineData(typeof(DictionaryWrapper), JsonTypeInfoKind.Dictionary)]
+        [InlineData(typeof(List<int>), JsonTypeInfoKind.Enumerable)]
+        [InlineData(typeof(ListWrapper), JsonTypeInfoKind.Enumerable)]
+        [InlineData(typeof(int[]), JsonTypeInfoKind.Enumerable)]
+        public static void JsonTypeInfoKindIsReportedCorrectly(Type type, JsonTypeInfoKind expectedJsonTypeInfoKind)
```

Comment:
Could this be a generic theory?

Reply:
You can use a [witness type](https://github.com/dotnet/runtime/blob/71470821df251c43a7bd4a046cb36a11db27d06d/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Stream.Collections.cs#L26) to serve this purpose. Slightly artifical but still simpler than doing reflection invocation.

Reply:
I've just tried that but seem there is no way to create witness as a const and use it in custom attribute (I also tried changing that to struct and use `default` but with no success). Without that I need to change InlineData into MemberData and that makes it less readable.

Reply:
Yeah needs a MemberData source unfortunately.